gazebo_ros/include/gazebo_ros/qos.hpp

```diff
@@ -0,0 +1,202 @@
+   *   settings.
+   *   This contains the necessary information extracting the remappings.
+   * \throws gazebo_ros::InvalidQoSException if there is an invalid QoS value or a topic element
+   *   is missing a "name" attribute.
+   */
+  QoS(
+    sdf::ElementPtr _sdf,
+    const std::string node_name,
+    const std::string node_namespace,
+    const rclcpp::NodeOptions & options);
+
+  /// Get the QoS for a publisher
+  /*
+   * \param[in] topic: Get QoS for publishers on this topic name.
+   * \param[in] default_qos: The default quality of service used for settings that have not been
+   *   overridden.
+   */
+  rclcpp::QoS get_publisher_qos(
+    const std::string topic, rclcpp::QoS default_qos = rclcpp::QoS(10)) const;
+
+  /// Get the QoS for a subscription
+  /*
+   * \param[in] topic: Get QoS for subscriptions on this topic name.
+   * \param[in] default_qos: The default quality of service used for settings that have not been
+   *   overridden.
+   */
+  rclcpp::QoS get_subscription_qos(
+    const std::string topic, rclcpp::QoS default_qos = rclcpp::QoS(10)) const;
+
+private:
+  /// Storage container for QoS overrides for a single profile.
+  struct QoSOverrides
+  {
+    rmw_qos_reliability_policy_t reliability;
+    rmw_qos_durability_policy_t durability;
+    rmw_qos_history_policy_t history;
+    rmw_qos_liveliness_policy_t liveliness;
+    size_t depth;
+    std::chrono::milliseconds deadline;
+    std::chrono::milliseconds lifespan;
+    std::chrono::milliseconds liveliness_lease;
+
+    QoSOverrides()
+    : reliability(RMW_QOS_POLICY_RELIABILITY_UNKNOWN),
+      durability(RMW_QOS_POLICY_DURABILITY_UNKNOWN),
+      history(RMW_QOS_POLICY_HISTORY_UNKNOWN),
+      liveliness(RMW_QOS_POLICY_LIVELINESS_UNKNOWN),
+      depth(0),
+      deadline(0),
+      lifespan(0),
+      liveliness_lease(0)
+    {}
+  };
+
+  /// Helper function for parsing an SDF to get QoS overrides.
+  static QoSOverrides get_qos_overrides_from_sdf(sdf::ElementPtr _sdf);
+
+  /// Helper function for applying overrides on top of a default QoS profile.
+  static rclcpp::QoS apply_overrides(
+    const QoSOverrides & overrides, const rclcpp::QoS default_qos);
+
+  /// Get the remapped name for a topic
+  std::string get_remapped_topic_name(const std::string topic) const;
+
+  /// Map topic names to publisher QoS overrides
+  std::map<std::string, QoSOverrides> publisher_qos_overrides_map_;
+  /// Map topic names to subscription QoS overrides
+  std::map<std::string, QoSOverrides> subscription_qos_overrides_map_;
+
+  // The following members are needed for getting topic remappings
+
+  /// The name of the node associated with the QoS settings
+  std::string node_name_;
+
+  /// The namespace of the node associated with the QoS settings
+  std::string node_namespace_;
+
+  /// The options of the node associated with the QoS settings
+  rclcpp::NodeOptions node_options_;
```

Comment:
@chapulina I'm not sure if we're using the PIMPL pattern in `gazebo_ros_pkgs` yet, but I'm wondering if this new class is a good candidate for it, since it has lots of private member variables

Reply:
Thank you for catching this, @scpeters . Indeed we're using PIMPL for the ROS 2 packages, as described on the [Contribution guide](https://github.com/ros-simulation/gazebo_ros_pkgs/wiki/ROS-2-contribution-guide#in-the-code).